neonvm/runner/containermgr.go

```diff
@@ -0,0 +1,364 @@
+		err := server.Shutdown(context.Background())
+		logger.Info("shut down cpu_change server", zap.Error(err))
+	}
+}
+
+func (c *containerInfo) handleCPUChange(logger *zap.Logger, crictl *crictlConfig, w http.ResponseWriter, r *http.Request) {
+	if r.Method != "POST" {
+		logger.Error("unexpected method", zap.String("method", r.Method))
+		w.WriteHeader(400)
+		return
+	}
+	body, err := io.ReadAll(r.Body)
+	if err != nil {
+		logger.Error("could not read body", zap.Error(err))
+		w.WriteHeader(400)
+		return
+	}
+
+	var parsed api.VCPUChange
+	if err = json.Unmarshal(body, &parsed); err != nil {
+		logger.Error("could not parse body", zap.Error(err))
+		w.WriteHeader(400)
+		return
+	}
+
+	logger.Info("got CPU update", zap.Float64("CPU", parsed.VCPUs.AsFloat64()))
+
+	shares := sharesForCPU(parsed.VCPUs)
+	quota := int64(vmv1.MilliCPU(cpuLimitOvercommitFactor*parsed.VCPUs).AsFloat64() * float64(cpuPeriod))
+
+	logger.Info(
+		"calculated CPU quantities for vCPU",
+		zap.Int("shares", shares),
+		zap.Int64("quota", quota),
+		zap.Int("period", cpuPeriod),
+	)
+
+	// update container
+	err = crictl.run(
+		logger,
+		nil,
+		"update",
+		"--cpu-share", strconv.Itoa(shares),
+		"--cpu-quota", strconv.FormatInt(quota, 10),
+		"--cpu-period", strconv.Itoa(cpuPeriod),
+		c.containerID,
+	)
+	if err != nil {
+		logger.Error("could not update container CPU", zap.String("id", c.containerID), zap.Error(err))
+		w.WriteHeader(500)
+		return
+	}
+
+	// store the milli CPU now that we've set it, so that in handleCPUCurrent we can handle rounding
+	// issues based on the last operation we did.
+	c.lastMilliCPU.Store(uint32(parsed.VCPUs))
+
+	w.WriteHeader(200)
+}
+
+func (c *containerInfo) handleCPUCurrent(logger *zap.Logger, crictl *crictlConfig, w http.ResponseWriter, r *http.Request) {
+	if r.Method != "GET" {
+		logger.Error("unexpected method", zap.String("method", r.Method))
+		w.WriteHeader(400)
+		return
+	}
+
+	logger.Info("got CPU current request")
+
+	var inspect crictlContainerInspect
+	if err := crictl.run(logger, &inspect, "inspect", c.containerID); err != nil {
+		logger.Error("could not inspect container", zap.String("id", c.containerID), zap.Error(err))
+	}
+
+	shares := int(inspect.Info.RuntimeSpec.Linux.Resources.CPU.Shares)
+
+	logger.Info(
+		"fetched current CPU shares",
+		zap.Int("shares", shares),
+	)
+
+	last := vmv1.MilliCPU(c.lastMilliCPU.Load())
+	expectedIfNoChange := sharesForCPU(last)
+
+	var resp api.VCPUCgroup
+	if shares == expectedIfNoChange {
+		resp = api.VCPUCgroup{VCPUs: last}
+	} else {
+		resp = api.VCPUCgroup{VCPUs: cpuForShares(shares)}
+	}
+
+	logger.Info("responding with current CPU", zap.Float64("cpu", resp.VCPUs.AsFloat64()))
+
+	body, err := json.Marshal(resp)
+	if err != nil {
+		logger.Error("could not marshal body", zap.Error(err))
+		w.WriteHeader(500)
+		return
+	}
+
+	w.Header().Add("Content-Type", "application/json")
+	w.Write(body) //nolint:errcheck // Not much to do with the error here. TODO: log it?
```

Comment:
Let's log the error here. Better than ignoring it.

Reply:
Copied from neonvm-runner, happy to address in a follow-up

https://github.com/neondatabase/autoscaling/blob/32e407a6ba1ae49c341e9de898d62489914fff74/neonvm/runner/main.go#L771